devito/finite_differences/tools.py

```diff
@@ -169,20 +189,71 @@ def generate_indices(func, dim, order, side=None, x0=None):
 
 
 def generate_indices_cartesian(dim, order, side, x0):
+    """
+    Indices for the finite-difference scheme on a cartesian grid
+
+    Parameters
+    ----------
+    dim: Dimension
+        Dimensions w.r.t which the derivative is taken
+    order: Int
+        Order of the finite-difference scheme
+    side: Side
+        Side of the scheme, (centered, left, right)
+    x0: Dimension or Expr or Number
+        Origin of the scheme, ie. `x`, `x + .5 * x.spacing`, ...
+
+    Returns
+    -------
+    Ordered list of indices
+    """
     shift = 0
+    # Min/max offsets for indices
+    o_start = -order//2
+    o_end = order//2 + 1
+    # Shift if x0 is not on the grid
+    offset_c = 0 if sympify(x0).is_Integer else (dim - x0)/dim.spacing
```

Comment:
probably best with a try-catch because sympify is slow

we need to make sure these routines are as quick as possible

Reply:
Again, will be done once and only once

Reply:
And this is only done on `x0` that is a Symbol or a Number, not an expression.